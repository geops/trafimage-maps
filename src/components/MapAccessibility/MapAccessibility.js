import OLMap from 'ol/Map';
import PropTypes from 'prop-types';
import { useEffect } from 'react';
import { Layer } from 'mobility-toolbox-js/src/ol/';

const KEYCODE_TAB = 9;
const KEYCODE_ENTER = 13;

/**
 * This component provides accessibility for map features.
 *
 * Accessibility needs to be enabled for top-level layers
 * by setting the hasAccessibility property. For now only
 * single (radioGroup) TrackerLayers are supported.
 */
const MapAccessibility = ({ layers, map }) => {
  useEffect(() => {
    const mapTarget = map.getTarget();
    let tabFeature;
    let tabLayer;
    let tabFeatureIndex = 0;

    const mapAccessibility = (e) => {
      // cycle through map features
      if (e.which === KEYCODE_TAB) {
        [tabLayer] = layers
          .filter((l) => l.visible && l.properties.hasAccessibility)
          .map((l) => l.getVisibleChildren())
          .flat();
        if (mapTarget === document.activeElement && tabLayer) {
          if (tabLayer.isTrackerLayer) {
            const trajectories = tabLayer.tracker.getRenderedTrajectories();
            trajectories.sort((a, b) =>
              a.coordinate && a.coordinate[0] < b.coordinate[0] ? -1 : 1,
            );
            tabFeature = trajectories[tabFeatureIndex];

            if (tabFeature) {
              tabLayer.tracker.setHoverVehicleId(tabFeature.id);
              tabFeatureIndex += e.shiftKey ? -1 : 1;
              e.preventDefault();
            } else {
              tabLayer.tracker.setHoverVehicleId(null);
              tabFeature = null;
              tabFeatureIndex = 0;
            }
          }
        }
      }

      // click on highlighted feature
      if (e.which === KEYCODE_ENTER && tabFeature && tabLayer) {
        if (tabLayer.isTrackerLayer) {
          const { coordinate } = tabFeature;
          map.dispatchEvent({ type: 'singleclick', map, coordinate });
          e.preventDefault();
        }
      }
    };

    document.addEventListener('keydown', mapAccessibility);

    return function cleanup() {
      document.removeEventListener('keydown', mapAccessibility);
    };
  });

  return null;
};

MapAccessibility.propTypes = {
  layers: PropTypes.arrayOf(PropTypes.instanceOf(Layer)).isRequired,
  map: PropTypes.instanceOf(OLMap).isRequired,
};

export default MapAccessibility;
