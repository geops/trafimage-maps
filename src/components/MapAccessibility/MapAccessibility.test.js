import React from 'react';
import { mount } from 'enzyme';
import { Layer, TrackerLayer } from 'mobility-toolbox-js/ol';
import { Map, View } from 'ol';
import renderer from 'react-test-renderer';
import MapAccessibility from './MapAccessibility';

let map;
let mapElement;
const trackerLayer = new TrackerLayer({
  visible: true,
});
const layer = new Layer({
  visible: true,
  properties: { hasAccessibility: true },
  children: [trackerLayer],
});

describe('MapAccessibility', () => {
  beforeEach(() => {
    TrackerLayer.prototype.hasFeatureInfoAtCoordinate = jest.fn(() => true);
    mapElement = document.createElement('div');
    mapElement.tabIndex = 1;
    mapElement.tabindex = 1;
    document.body.appendChild(mapElement);
    map = new Map({
      target: mapElement,
      view: new View({
        center: [0, 0],
        zoom: 0,
      }),
    });
    trackerLayer.init(map);
    trackerLayer.tracker.renderedTrajectories = [
      { id: 1, coordinate: [1, 0] },
      { id: 2, coordinate: [0, 1] },
    ];
  });

  afterEach(() => {
    TrackerLayer.prototype.hasFeatureInfoAtCoordinate.mockRestore();
    trackerLayer.terminate(map);
    document.body.removeChild(mapElement);
  });

  test('should return null', () => {
    const component = renderer.create(
      <MapAccessibility map={map} layers={[layer]} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should go through vehicles on TAB event', () => {
    const event = new KeyboardEvent('keydown', { which: 9 });
    const eventShift = new KeyboardEvent('keydown', {
      which: 9,
      shiftKey: true,
    });
    const spy = jest.spyOn(event, 'preventDefault');
    const spy2 = jest.spyOn(eventShift, 'preventDefault');
    mount(<MapAccessibility map={map} layers={[layer]} />);
    mapElement.focus();
    // Go to vehicle with smaller x coordinate
    document.dispatchEvent(event);
    expect(trackerLayer.hoverVehicleId).toBe(2);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy2).toHaveBeenCalledTimes(0);
    spy.mockReset();
    spy2.mockReset();

    // Go to next vehicle
    document.dispatchEvent(event);
    expect(trackerLayer.hoverVehicleId).toBe(1);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy2).toHaveBeenCalledTimes(0);
    spy.mockReset();
    spy2.mockReset();

    // Go to previous vehicle
    document.dispatchEvent(eventShift);
    expect(trackerLayer.hoverVehicleId).toBe(2);
    expect(spy).toHaveBeenCalledTimes(0);
    expect(spy2).toHaveBeenCalledTimes(1);
    spy.mockReset();
    spy2.mockReset();

    // Go outside
    document.dispatchEvent(event);
    document.dispatchEvent(event);
    expect(trackerLayer.hoverVehicleId).toBe(null);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy2).toHaveBeenCalledTimes(0);
    spy.mockRestore();
    spy2.mockRestore();
  });

  test('should trigger maps singleclick event if a vehicle is selected', () => {
    const event = new KeyboardEvent('keydown', { which: 9 });
    const eventEnter = new KeyboardEvent('keydown', { which: 13 });
    const spy = jest.spyOn(eventEnter, 'preventDefault');
    const spy2 = jest.spyOn(map, 'dispatchEvent');
    mount(<MapAccessibility map={map} layers={[layer]} />);
    mapElement.focus();
    // Go to vehicle with smaller x coordinate
    document.dispatchEvent(event);
    document.dispatchEvent(eventEnter);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy2).toHaveBeenCalledTimes(1);
    expect(spy2.mock.calls[0][0].type).toBe('singleclick');
    expect(spy2.mock.calls[0][0].map).toBe(map);
    expect(spy2.mock.calls[0][0].coordinate).toEqual([0, 1]);
  });

  test('remove listener on unmount', () => {
    const event = new KeyboardEvent('keydown', { which: 9 });
    const spy = jest.spyOn(event, 'preventDefault');
    const wrapper = mount(<MapAccessibility map={map} layers={[layer]} />);
    wrapper.unmount();
    mapElement.focus();
    // Go to vehicle with smaller x coordinate
    document.dispatchEvent(event);
    expect(spy).toHaveBeenCalledTimes(0);
  });
});
