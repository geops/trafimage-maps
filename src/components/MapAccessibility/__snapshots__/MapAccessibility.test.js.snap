// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapAccessibility should return null 1`] = `null`;
